DOCS 942: Update Navigation Service docs for Python SDK support and add GetObstacles() (#1565)
viamrobotics/docs: docs/services/navigation/_index.md

The Navigation service is the stateful definition of Viam's [Motion Service](/services/motion/).
It uses GPS to autonomously navigate a rover [base](/components/base/) to user defined endpoints called `Waypoints`.

> Once these waypoints are added and the mode of the service is [set to `Mode.WAYPOINT`](#setmode), the service begins to define the robot's path.

Reviewer: ```suggestion
Once these waypoints are added and the mode of the service is [set to `MODE_WAYPOINT`](#setmode), the service begins to define the robot's path.
```

## Configuration

You must configure a [base](/components/base/) with a [movement sensor](/components/movement-sensor/) as part of your robot to configure a Navigation service.

> The movement sensor must support `GetPosition`, `GetCompassHeading`, or `Orientation` in its {{< glossary_tooltip term_id="model" text="model's" >}} implementation of the [Sensor API](/components/sensor/#api).

Author: @npentrel hmmm-- this is translated from a direct ask that I'm not sure if I am interpreting correctly: `Add: The movement sensor must support GetPosition and GetCompassHeading or Orientation. If your movement sensor can support both, then the motion planner will default to using the compass heading values`

@tessavitabile Is the following suggestion what your team means by that? If so, I think that would clear up Naomi's question.
And maybe I could provide more info on why this requirement is needed (if it needs combo of two sensor reading types or a GeoPoint)? 

```suggestion
The movement sensor must support both `GetPosition` and `GetCompassHeading`, or just `Orientation`, in its {{< glossary_tooltip term_id="model" text="model's" >}} implementation of the [Sensor API](/components/sensor/#api).
```

Reviewer 2: The movement sensor must support GetPosition, in order to know the robot's current GPS location. It must also support either GetCompassHeading or GetOrientation, in order to know which way the robot is facing. If the movement sensor supports both GetCompassHeading and GetOrientation, then GetCompassHeading will be used.

Author: @tessavitabile Perfect, thank you

> The movement sensor must support `GetPosition`, `GetCompassHeading`, or `Orientation` in its {{< glossary_tooltip term_id="model" text="model's" >}} implementation of the [Sensor API](/components/sensor/#api).

Author: ```suggestion
The movement sensor must support [`GetPosition()`](/components/movement-sensor/#getposition), and at least one of [`GetCompassHeading()`](/components/movement-sensor/#getcompassheading) or [`GetOrientation()`](/components/movement-sensor/#getorientation), in its {{< glossary_tooltip term_id="model" text="model's" >}} implementation of the [Movement Sensor API](/components/movement-sensor/#api).
It must support `GetPosition()` to report the robot's current GPS location.
It must also support either `GetCompassHeading()` or `GetOrientation()` to report which way the robot is facing.
```

> If your movement sensor provides multiple methods, your robot will default to using the values returned by `GetCompassHeading`.

Reviewer 3: and if it provides getposition and orientation?

        "store": {
            "type": "mongodb",
            "config": {
                "uri": "mongodb://127.0.0.1:12345"
            }
        }
    },
    "movement_sensor": "your-movement-sensor",
    "base": "your-base",
    "obstacles":[{
        "geometries": [{
            "label":"your-label-for-this-obstacle",
            "orientation":{
                "type":"ov_degrees",
                    "value":{
                        "x":1,
                        "y":0,
                        "z":0,
                        "th": 90
                    }
            },
            "x":10,
            "y":10,
            "z":10,

>             "translation":{

Reviewer 2: 'translation' is not supported for obstacles.

Author: Addressed

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `store` | obj | **Required** | The type and configuration of data storage to use. Either type `"memory"`, where no additional configuration is needed and the waypoints are stored in local memory while the Navigation process is running, or `"mongodb"`, where data persists at the specified [MongoDB URI](https://www.mongodb.com/docs/manual/reference/connection-string) of your MongoDB deployment. |

> | `base` | string | **Required** | The `name` you have configured for the [base](/components/base/) you are operating with this service. |

Reviewer: Nice catch!

> ### Mode

Reviewer: Hmm, perhaps for engineering: I think `Mode` as get() function is misleading, feels conceptual when talking about both get() and set() tasks. I might suggest `getMode` for the fetching to match `setMode` for the setting.

No action in docs -- exact match to literal function nomenclature is all we can do! -- but I'll consider an eng ticket or feedback or such. Do you feel the same or am I totally off-base here?

Author: def an ask for eng but a worthy one-- Most of the get endpoints have been named in this convention, so I assumed it was what they wanted with Go.

Reviewer: Thanks! I poked around and indeed this appear to follow our existing go naming conventions, so I'll leave it alone. Thanks for the helpful context!

Get the `Mode` the service is operating in.

There are three options for modes: `MODE_UNSPECIFIED`, `MODE_MANUAL`, or `MODE_WAYPOINT`.

> - `MODE_UNSPECIFIED`: Do not perform any mode of autonomous navigation.

Reviewer: Question re: `any mode`: Is it possible to purposefully set `MODE_UNSPECIFIED` myself, or does it only appear for a newly-created service until otherwise configured? If the former, `MODE_UNSPECIFIED` technically becomes a `Mode` itself, for example: as returned by `setMode`. If the latter, than this language is correct.

### Mode

Get the `Mode` the service is operating in.

There are three options for modes: `MODE_UNSPECIFIED`, `MODE_MANUAL`, or `MODE_WAYPOINT`.

> - `MODE_UNSPECIFIED`: Do not perform autonomous navigation.

Reviewer 2: The default is MODE_MANUAL. MODE_UNSPECIFIED is just in proto, since all enums have to have an unspecified option. We don't need to document MODE_UNSPECIFIED.

Author: Ok, that makes sense. Have addressed!

```go
myNav, err := navigation.FromRobot(robot, "my_nav_service")

// Get the Mode the service is operating in
mode, err := myNav.Mode(context.Background(), nil)
```

{{% /tab %}}
{{% tab name="Python" %}}

**Parameters:**

- `timeout` [(Optional\[float\])](https://docs.python.org/library/typing.html#typing.Optional): An option to set how long to wait (in seconds) before calling a time-out and closing the underlying RPC call.

**Returns:**

- [navigation.Mode.ValueType](https://python.viam.dev/autoapi/viam/services/navigation/index.html#viam.services.navigation.Mode): The `Mode` the service is operating in.

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/services/navigation/index.html#viam.services.navigation.NavigationClient.get_mode).

``` python
my_nav = NavigationClient.from_robot(robot=robot, name="my_nav_service")

# Get the Mode the service is operating in
await my_nav.get_mode()
```

{{% /tab %}}
{{< /tabs >}}

### SetMode

Set the `Mode` the service is operating in.

There are three options for modes: `MODE_UNSPECIFIED`, `MODE_MANUAL`, or `MODE_WAYPOINT`.

> - `MODE_UNSPECIFIED`: Do not perform any mode of autonomous navigation.

Reviewer: Same comment as line 148.

Author: done!

> - `MODE_UNSPECIFIED`: Do not perform any mode of autonomous navigation.

Author: ```suggestion
- `MODE_UNSPECIFIED`: Do not perform any autonomous navigation.
```

> - [(*geo.Point)](https://pkg.go.dev/github.com/kellydunn/golang-geo#Point): The current location of the robot in the navigation service, represented in a `Point` with latitude (lat) and longitude (lng) values.

Reviewer: Same comment as 290 below.

> - [(*geo.Point)](https://pkg.go.dev/github.com/kellydunn/golang-geo#Point): The current location of the robot in the navigation service, represented in a `Point` with latitude (lat) and longitude (lng) values.

Author: ```suggestion
- [(*geo.Point)](https://pkg.go.dev/github.com/kellydunn/golang-geo#Point): The current location of the robot in the navigation service, represented in a `Point` with latitude and longitude values.
```

- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk/services/navigation#Service).

```go
myNav, err := navigation.FromRobot(robot, "my_nav_service")

// Get the current location of the robot in the navigation service
location, err := myNav.Location(context.Background(), nil)
```

{{% /tab %}}
{{% tab name="Python" %}}

**Parameters:**

- `timeout` [(Optional\[float\])](https://docs.python.org/library/typing.html#typing.Optional): An option to set how long to wait (in seconds) before calling a time-out and closing the underlying RPC call.

**Returns:**

> - [(navigation.GeoPoint)](https://python.viam.dev/autoapi/viam/services/navigation/index.html#viam.services.navigation.GeoPoint): The current location of the robot in the navigation service, represented in a `GeoPoint` with latitude (lat) and longitude (lng) values.

Reviewer: Hmm, where did you get `lat` and `lng` from? They appear to be the full words according to the linked SDK reference entry. 

EDIT: Ah, I see earlier references to these names, betting they changed at some point separate from this code change.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk/services/navigation#Service).

```go
myNav, err := navigation.FromRobot(robot, "my_nav_service")

// Get an array containing each waypoint stored by the navigation service
waypoints, err := myNav.Waypoints(context.Background(), nil)
```

{{% /tab %}}
{{% tab name="Python" %}}

**Parameters:**

- `timeout` [(Optional\[float\])](https://docs.python.org/library/typing.html#typing.Optional): An option to set how long to wait (in seconds) before calling a time-out and closing the underlying RPC call.

**Returns:**

- [(List[navigation.Waypoint])](https://python.viam.dev/autoapi/viam/services/navigation/index.html#viam.services.navigation.Waypoint): An array comprised of each `Waypoint` in the service's data storage.
These are locations designated within a path for the robot to navigate to.

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/services/navigation/index.html#viam.services.navigation.NavigationClient.get_waypoints).

```python
my_nav = NavigationClient.from_robot(robot=robot, name="my_nav_service")

> # Get an list containing each waypoint stored by the navigation service

Reviewer: ```suggestion
# Get a list containing each waypoint stored by the navigation service
```

err := myNav.AddWaypoint(context.Background(), location, nil)
```

{{% /tab %}}
{{% tab name="Python" %}}

**Parameters:**

> - `point`[(navigation.GeoPoint)](https://python.viam.dev/autoapi/viam/services/navigation/index.html#viam.services.navigation.GeoPoint): The current location of the robot in the navigation service, represented in a `GeoPoint` with latitude (lat) and longitude (lng) values.

Reviewer: Same comment as line 290

Author: Thank you, should have addressed!
